### PR TITLE
Node execution engine and build tooling

### PR DESCRIPTION
**Pull request by sandorfr**
Addresses #909

### CODE REVIEW

**Comment by sandorfr** [comment]
There is still a few things to be done especially (https://github.com/geeklearningio/gl-vsts-tasks-build-scripts/issues/1), but I'm creating the PR so progress can be followed.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+var additionalArguments = tl.getInput('additionalArguments');
+var gitVersionPath = tl.getInput('gitVersionPath');
+
+var currentDirectory = __dirname;
+
+var sourcesDirectory = tl.getVariable("Build.SourcesDirectory")
+
+if (!gitVersionPath) {
+    gitVersionPath = tl.which("GitVersion.exe");
+    gitVersionPath = path.join(currentDirectory, "GitVersion.exe");
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: This doesn't make sense. I assume the second assignment should be the fallback in case `which` couldn't find the tool?

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        var toolRunner = tl.tool(gitVersionPath);
+
+        toolRunner.arg([
+            sourcesDirectory,
+            "/output",
+            "buildserver",
+            "/nofetch"
+        ]);
+
+
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Remove second blank line

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        if (additionalArguments) {
+            toolRunner.arg(additionalArguments);
+        }
+
+        var result = await toolRunner.exec();
+        if (result) {
+            tl.setResult(tl.TaskResult.Failed, "An error occured during GitVersion execution")
+        } else {
+
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Remove blank line

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            // workaround gitversion to make sure variable are availabel in all vsts build contexts
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Typo: Available

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            tl.setResult(tl.TaskResult.Succeeded, "GitVersion executed successfully")
+        }
+    }
+    catch (err) {
+        tl.debug(err.stack)
+        tl.setResult(tl.TaskResult.Failed, err);
+    }
+})();
+
+
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Remove blank lines

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 4 above)
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: What's the issue which should be worked around here? Passing variables to the build system is something which should be handled inside of GitVersion itself, not as part of this task.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+var currentDirectory = __dirname;
+
+var sourcesDirectory = tl.getVariable("Build.SourcesDirectory")
+
+if (!gitVersionPath) {
+    gitVersionPath = tl.which("GitVersion.exe");
+    gitVersionPath = path.join(currentDirectory, "GitVersion.exe");
+}
+
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Should we add an additional check here if GitVersion.exe could be found. Or is this case already nicely handled by the tool runner?

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
+import tl = require('vsts-task-lib/task');
+import path = require('path');
+import fs = require('fs-extra');
+import q = require('q');
+import XRegExp = require('xregexp');
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Seems not to be used

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import tl = require('vsts-task-lib/task');
+import path = require('path');
+import fs = require('fs-extra');
+import q = require('q');
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Seems not to be used

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import tl = require('vsts-task-lib/task');
+import path = require('path');
+import fs = require('fs-extra');
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Seems not to be used

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
+import tl = require('vsts-task-lib/task');
+import path = require('path');
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Seems not to be used

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  "private": true,
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "fs-extra": "0.30.0",
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: Seems not to be used

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    "vsts-task-lib": "0.9.20",
+    "q" : "1.4.1"
```
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: Seems not to be used

File: .gitmodules [hunk 13]
```diff
@@ -0,0 +1,3 @@
+[submodule "src/GitVersionTfsTask/BuildScripts"]
```
Comment (pascalberger) [inline .gitmodules]: As already mentioned instead of a submodule this should be included by npm (see https://github.com/geeklearningio/gl-vsts-tasks-build-scripts/issues/1).

The whole build pipeline needs generally a little bit more work. It needs to be integrated into the existing build pipeline.

File: src/GitVersionTfsTask/typings.json [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/typings.json]: I would prefer we use typescript 2 and `@types` rather than typings

File: src/GitVersionTfsTask/package.json [hunk 15]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "git-version-tfs-task",
+  "version": "1.0.0",
+  "description": "",
+  "private": true,
+  "main": "index.js",
+  "author": "",
+  "license": "ISC",
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/package.json]: Should use MIT to match GitVersion

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 10 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: tl and path are used almost everywhere. ;)

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 9 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: fs-extra is actually not used, I'll remove the dep

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 8 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: I've had some issue with q not being properly loaded (vsts task lib relies on it) in the past. I'll see if it can be removed.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 7 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: xregexp will be removed.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 1 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: There is an if missing indeed.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 6 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: The process would be try to find gitversion.exe in path. If not found use bundled version. I think adding a advanced option `Always use bundled version` could be nice too.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 4 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: Correct me if I'm wrong but currently GitVersion justs sets environment variable. It should be set with a proper [set command](https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md). I faced a few case where gitversion where not properly expanded because of that in the past. Does it belong in gitversion (using a specific arguement) or in the task I leave it up to you. (This is the reason we are always checking for unexpanded variables in our file patch extension).

File: .gitmodules (hunk 13 above)
Comment (sandorfr) [inline .gitmodules]: I'll focus on fixing everything on the task itself. Once we have something working and satisfying and depending on your preference I will either publish our tooling on npm or drop our tooling from the PR. Meanwhile it allows me to publish private extension for testing purpose more easily.

File: src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts (hunk 4 above)
Comment (pascalberger) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: @sandorfr GitVersion currently already uses logging commands, see [here](https://github.com/GitTools/GitVersion/blob/master/src/GitVersionCore/BuildServers/VsoAgent.cs#L18)

Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/GitVersion.ts]: ok. I'll see if I can still reproduce the issue (It was last year though)

File: .gitmodules (hunk 13 above)
Comment (sandorfr) [inline .gitmodules]: We've decided to go forward with the npm package. So I've updated this PR accordingly. I'll investigate the vsts beta feature.

File: src/GitVersionTfsTask/Configuration.json [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+            "VssExtensionGalleryFlags": [
+                "Public"
+            ],
+            "DisplayNamesSuffix": "",
+            "TaskIds": {
+                "GitVersion" : "e5983830-3f75-11e5-82ed-81492570a08e"
+            }
+        },
+        {
+            "Name": "dev",
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Configuration.json]: Is this a requirement, I would prefer `prerelease` or something?

File: src/GitVersionTfsTask/Extension/vss-extension.json [hunk 17]
```diff
@@ -1,9 +1,8 @@
 {
     "manifestVersion": 1,
-    "id": "gitversion",
+    "id": "gl-gitversion",
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: Please revert this change

File: src/GitVersionTfsTask/Extension/vss-extension.json [hunk 18]
```diff
@@ -1,9 +1,8 @@
...
     "name": "GitVersion",
-    "publisher": "gittools",
-    "public": true,
+    "publisher": "geeklearningio",
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: And this one

File: src/GitVersionTfsTask/Extension/vss-extension.json [hunk 19]
```diff
@@ -1,9 +1,8 @@
 {
     "manifestVersion": 1,
-    "id": "gitversion",
+    "id": "gl-gitversion",
     "name": "GitVersion",
-    "publisher": "gittools",
-    "public": true,
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: Why remove `public`?

File: src/GitVersionTfsTask/Extension/vss-extension.json [hunk 20]
```diff
@@ -46,27 +45,15 @@
...
-            "path": "img/build-task.png"
+            "path": "build-task.png"
         }
     ],
     "content": {
         "details": {
             "path": "overview.md"
         }
-    },
-    "contributions": [
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: Have they moved more towards convention over configuration?

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json [hunk 21]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "git-version",
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: remove the `-`, should be `gitversion`

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+  "version": "1.0.0",
+  "private": true,
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: Why private?

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "MIT",
+  "dependencies": {
+    "vsts-task-lib": "0.9.20",
+    "q" : "1.4.1"
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: Why not native promises?

File: src/GitVersionTfsTask/tsconfig.json [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+        "target": "es6",
+        "sourceMap": false,
+        "inlineSourceMap": true,
+        "declaration": true
+    },
+    "filesGlob": [
+        "./Tasks/*.ts",
+        "./Tests/*.ts",
+        "./typings/index.d.ts",
+        "./Commom/Node/*.d.ts"
```
Comment (JakeGinnivan) [inline src/GitVersionTfsTask/tsconfig.json]: Spelling, should be `Common`?

File: src/GitVersionTfsTask/Extension/vss-extension.json (hunk 17 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: I'll revert it when it's ready, meanwhile it allows me to publish a prerelease version of it without any conflicts :)

File: src/GitVersionTfsTask/Configuration.json (hunk 16 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Configuration.json]: we can change it or remove it.

File: src/GitVersionTfsTask/Extension/vss-extension.json (hunk 20 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: no this just filled automatically by our tooling, if you decided to keep [the tooling](https://www.npmjs.com/package/gl-vsts-tasks-build-scripts), they are no longer needed.

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json (hunk 22 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: this is not meant to be published to npm, I allows some of the usually fields required by package.json to be missing.

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json (hunk 23 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: vsts-task-lib depends on q. I guess this is related to the fact that some agents are running older versions of node which won't provide native Promises.

File: src/GitVersionTfsTask/Tasks/GitVersion/package.json (hunk 21 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Tasks/GitVersion/package.json]: changing it but it does not really matter as this file is just here to indicate the dependency to be bundled with the task.

File: src/GitVersionTfsTask/Extension/vss-extension.json (hunk 19 above)
Comment (sandorfr) [inline src/GitVersionTfsTask/Extension/vss-extension.json]: Our tooling allows you to build different version of the same extension. In order to do that we use the extension flags instead and they are configured in `configuration.json`.

**Comment by sandorfr** [comment]
Hi @JakeGinnivan @pascalberger, thanks for taking the time to review this. Just mentioning if it was not clear that this PR is more to show the work in progress, and comment on the core of it (`gitversion.ts`). 

I know that our tooling might be confusing for you, but I need it to be able to use our continuous delivery pipeline which I use to actually build publish and test this extension. But when my work will be done, and the tasks will be 100% functional, I can make a new PR without any of our tooling. We think it's great but don't want to be pushy about it.

Meanwhile the real issue I'm facing is that the current portable version crashes on the Mac. The good news is this is probably a gitversion bug which has been fixed in https://github.com/libgit2/libgit2sharp/pull/1409. The bad news is a version of gitversion including this fix hasn't been released yet :(

**Comment by JakeGinnivan** [comment]
I would love to get this in for personal reasons :)

Would we be able to drop the initial cut without the shared tooling then switch over to the reusable tooling once it is ready?

**Comment by JakeGinnivan** [comment]
Once we merge this into master I will release it soon after that and you will not need your own GitVersion pipeline

**Comment by JakeGinnivan** [comment]
Would also allow us to release pre-release packages to VSTS and also close issues like #1150

**Comment by sandorfr** [comment]
Ok, @JakeGinnivan I try to find some time to make the new minimal PR very soon.

libgit2sharp apparently released a pre-release with the fix: https://www.nuget.org/packages/LibGit2Sharp/0.24.0-pre20170124092542. If someone can see on your side if there is a way for you to update the lib so it really works on macosx, that would be awesome :)
